File: algo-03/art-optimal-linkedlist.py
'''
root@raspberrypi:~/projet-smb215# python3 algo-03/art-optimal.py
root@raspberrypi:~/projet-smb215# perf stat python3 algo-03/art-optimal.py
'''

import os


DIR = os.path.dirname(os.path.abspath(__file__))
INPUT_FILE = DIR + "/input.txt"
OUTPUT_FILE = DIR + "/solution.txt"


class DrawableElem:
    def __init__(self, x, y, d, w, i, prev=None, next=None):
        self.x = x
        self.y = y
        self.d = d
        self.w = w
        self.i = i
        self.prev = prev
        self.next = next


def getArrayIndex(x, y):
    return y * baseX + x


def inDrawableArea(x, y, d):
    return x + d < baseX and y + d < baseY


def isDrawable(x, y, d):
    if baseGrid[y + d][x + d] == "*":
        return False
    for i in range(d):
        if baseGrid[y + i][x + d] == "*" or baseGrid[y + d][x + i] == "*":
            return False
    return True


def compareElem(a, b):
    if a.w == b.w:
        if a.d == b.d:
            if a.y == b.y:
                return a.x - b.x
            return a.y - b.y
        return b.d - a.d
    return b.w - a.w


def removeElem(elem):
    prev = elem.prev
    if prev is not None:
        prev.next = elem.next
        if elem.next is not None:
            elem.next.prev = prev
    else:
        drawableArea = elem.next
        drawableArea.prev = None
    return prev


def getIsolationIndex(x, y):
    index = 0
    for j in [-1, 0, 1]:
        yj = y + j
        if yj < 0 or yj > baseY - 1:
            continue
        for i in [-1, 0, 1]:
            xi = x + i
            if xi < 0 or xi > baseX - 1:
                continue
            if baseGrid[yj][xi] == "*":
                index += 1
    if index == 3:
        if 0 < y < baseY - 2:
            if 0 < x < baseX - 2:
                if baseGrid[y - 1][x] == "*" and baseGrid[y - 1][x - 1] == "*" and baseGrid[y - 1][x + 1] == "*":
                    return 0
                if baseGrid[y + 1][x] == "*" and baseGrid[y + 1][x - 1] == "*" and baseGrid[y + 1][x + 1] == "*":
                    return 0
                if baseGrid[y][x - 1] == "*" and baseGrid[y - 1][x - 1] == "*" and baseGrid[y + 1][x - 1] == "*":
                    return 0
                if baseGrid[y][x + 1] == "*" and baseGrid[y - 1][x + 1] == "*" and baseGrid[y + 1][x + 1] == "*":
                    return 0
                return index
    if index == 2:
        if 0 < y < baseY - 2:
            if 0 < x < baseX - 2:
                if baseGrid[y - 1][x] == "*" and (baseGrid[y - 1][x - 1] == "*" or baseGrid[y - 1][x + 1] == "*"):
                    return 0
                if baseGrid[y + 1][x] == "*" and (baseGrid[y + 1][x - 1] == "*" or baseGrid[y + 1][x + 1] == "*"):
                    return 0
                if baseGrid[y][x - 1] == "*" and (baseGrid[y - 1][x - 1] == "*" or baseGrid[y + 1][x - 1] == "*"):
                    return 0
                if baseGrid[y][x + 1] == "*" and (baseGrid[y - 1][x + 1] == "*" or baseGrid[y + 1][x + 1] == "*"):
                    return 0
                return index
    return index


def getDrawableArea():
    head = None
    prev = None
    for y in range(baseY):
        for x in range(baseX):
            if baseGrid[y][x] == "#":
                d = 1
                while True:
                    if not inDrawableArea(x, y, d) or not isDrawable(x, y, d):
                        break
                    d += 1
                elem = DrawableElem(x, y, d, d ** 2, getIsolationIndex(x, y), prev)
                if prev is None:
                    head = elem
                else:
                    prev.next = elem
                prev = elem
    return head


def isInArea(x, y, elem):
    return elem.x <= x < elem.x + elem.d and elem.y <= y < elem.y + elem.d


def computeWeightArea(x, y, d):
    weight = 0
    for j in range(d):
        for i in range(d):
            if baseGrid[y + j][x + i] == "#":
                weight += 1
    return weight


def fillHeaviestArea(x, y):
    found = None
    elem = drawableArea
    while elem.next is not None:
        if isInArea(x, y, elem):
            elem.w = computeWeightArea(elem.x, elem.y, elem.d)
            if elem.w == 0:
                removeElem(elem)
            else:
                if found is None:
                    found = elem
                elif compareElem(elem, found) < 0:
                    found = elem
        elem = elem.next
    if found is not None and found.w > 0:
        addToResult(found.x, found.y, found.d)
        return removeElem(found)
    return None


def fillGrid(x, y, d):
    for j in range(d):
        for i in range(d):
            if baseGrid[y + j][x + i] == "#":
                baseGrid[y + j][x + i] = "+"


def addToResult(x, y, d):
    # print(f'FILL,{x},{y},{d}')
    result.append(f'FILL,{x},{y},{d}')
    return fillGrid(x, y, d)


result = []
baseGrid = []
drawableArea = None

with open(INPUT_FILE) as f:
    baseX, baseY = map(int, f.readline().strip().split(','))
    for line in f:
        pixels = list(line.strip())
        baseGrid.append(pixels)
        # isolationIndex.append([0] * baseX)

# print(f'BaseX: {baseX}, BaseY: {baseY}')
drawableArea = getDrawableArea()

# Fill in the most isolated pixels by heaviest area
for index in range(8, 1, -1):
    elem = drawableArea
    while elem.next is not None:
        if baseGrid[elem.y][elem.x] == "#" and index == elem.i:
            elem = fillHeaviestArea(elem.x, elem.y)
        if elem is not None and elem.next is not None:
            elem = elem.next
        else:
            elem = drawableArea

# Fill in the remaining pixels from the end line
for y in range(baseY - 1, -1, -1):
    for x in range(baseX):
        if baseGrid[y][x] == "#":
            fillHeaviestArea(x, y)

with open(OUTPUT_FILE, 'w') as f:
    f.write('\n'.join(result))
print('Res:', len(result))

File: algo-02/art-optimal-linkedlist.c
/*
root@raspberrypi:~/projet-smb215# gcc -w -Wall algo-02/art-optimal.c -o algo-02/art-optimal.o
root@raspberrypi:~/projet-smb215# chmod +x algo-02/art-optimal.o
root@raspberrypi:~/projet-smb215# perf stat ./algo-02/art-optimal.o
*/

#include <errno.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define LONG_MAX     2147483647L
#define LONG_MIN (-LONG_MAX - 1L)

#define INPUT_FILE  "./algo-02/input.txt"
#define OUTPUT_FILE "./algo-02/solution.txt"

#define HASH '\043'
#define STAR '\052'
#define PLUS '\053'

typedef struct drawableElem {
    int x; // coord X
    int y; // coord Y
    int d; // deep max
    int w; // weight
    struct drawableElem *prev;
    struct drawableElem *next;
} drawableElem;

static FILE *inFptr = NULL;
static FILE *outFptr = NULL;
static int result = 0;
static int baseX = 0; // num of rows
static int baseY = 0; // num of cols (line length)
static int gridLen = 0;
static char *baseGrid;
static drawableElem *drawableArea;

int parseInt(char *str, int *value) {
    char *endptr;
    errno = 0; // to distinguish success/failure after call
    long int lvalue = strtol(str, &endptr, 10);
    if ((errno == ERANGE && (lvalue == LONG_MAX || lvalue == LONG_MIN)) || (errno != 0 && lvalue == 0)) {
        printf("Conversion Error !\n");
        return -1;
    }
    *value = (int) lvalue;
    return 0;
}

int parseFirstLine(FILE *inFp) {
    int rc = 0;
    char line[32] = {0};
    if (fgets(line, sizeof(line), inFp) == NULL) {
        rc = -1;
        printf("Unable to get the first line: %s\n", strerror(errno));
        goto __error;
    }
    char *baseX_str = strtok(line, ",");
    char *baseY_str = strtok(NULL, ",");
    if ((rc = parseInt(baseX_str, &baseX)) < 0) {
        goto __error;
    }
    if ((rc = parseInt(baseY_str, &baseY)) < 0) {
        goto __error;
    }
    __error:
    return rc;
}

int computeBaseGrid(FILE *inFp) {
    char *baseGrid_pos0 = baseGrid;
    int len = baseX + 4; // add 4 bytes for the end of the line and end of the string
    char line[len];
    memset(line, 0, sizeof(line));
    // Lines are placed one after the other in the memory
    // baseGrid[0][0] is pos0
    // baseGrid[1][0] is pos0 + lineLength
    // To know the position of baseGrid[x][y], baseGrid = pos0 + y * baseX + x
    while (fgets(line, len, inFp) != NULL) {
        memcpy(baseGrid, line, baseX);
        memset(line, 0, sizeof(line));
        baseGrid += baseX; // move the pointer
    }
    // Back to position 0
    baseGrid = baseGrid_pos0;
    return 0;
}

int getArrayIndex(int x, int y) {
    return y * baseX + x;
}

char getBaseGrid(int x, int y) {
    return *(baseGrid + getArrayIndex(x, y));
}

bool isBetween(int value, int min, int max) {
    return value > min && value < max;
}

bool inDrawableArea(int x, int y, int d) {
    return x + d < baseX && y + d < baseY;
}

bool isDrawable(int x, int y, int d) {
    if (getBaseGrid(x + d, y + d) == STAR) {
        return false;
    }
    for (int i = 0; i < d; i++) {
        if (getBaseGrid(x + d, y + i) == STAR || getBaseGrid(x + i, y + d) == STAR) {
            return false;
        }
    }
    return true;
}

int compareElem(drawableElem a, drawableElem b) {
    if (a.w == b.w) {
        if (a.d == b.d) {
            if (a.y == b.y) {
                return a.x - b.x;
            }
            return a.y - b.y;
        }
        return b.d - a.d;
    }
    return b.w - a.w;
}

drawableElem *removeElem(drawableElem *elem) {
    drawableElem *prev = elem->prev;
    if (prev != NULL) {
        prev->next = elem->next;
        elem->next->prev = prev;
    } else {
        drawableArea = elem->next;
        drawableArea->prev = NULL;
    }
    free(elem);
    return prev;
}

drawableElem *getDrawableArea() {
    drawableElem *head = NULL;
    drawableElem *parent = NULL;
    for (int y = 0; y < baseY; y++) {
        for (int x = 0; x < baseX; x++) {
            if (getBaseGrid(x, y) == HASH) {
                int d = 1;
                while (inDrawableArea(x, y, d) && isDrawable(x, y, d)) {
                    d++;
                }
                drawableElem *elem = (drawableElem *) malloc(sizeof(drawableElem));
                elem->x = x;
                elem->y = y;
                elem->d = d;
                elem->w = d * d;
                elem->prev = parent;
                elem->next = NULL;
                if (parent == NULL) {
                    head = elem;
                } else {
                    parent->next = elem;
                }
                parent = elem;
            }
        }
    }
    return head;
}

bool isInArea(int x, int y, drawableElem elem) {
    return isBetween(x, elem.x - 1, elem.x + elem.d) && isBetween(y, elem.y - 1, elem.y + elem.d);
}

int computeWeightArea(drawableElem elem) {
    int weight = 0;
    for (int j = 0; j < elem.d; j++) {
        for (int i = 0; i < elem.d; i++) {
            if (getBaseGrid(elem.x + i, elem.y + j) == HASH) {
                weight++;
            }
        }
    }
    return weight;
}

void fillGrid(drawableElem elem) {
    for (int j = 0; j < elem.d; j++) {
        for (int i = 0; i < elem.d; i++) {
            if (getBaseGrid(elem.x + i, elem.y + j) == HASH) {
                int yxIndex = getArrayIndex(elem.x + i, elem.y + j);
                baseGrid[yxIndex] = PLUS;
            }
        }
    }
}

void addToResult(drawableElem elem) {
    // console.log(`FILL,${elem.x},${elem.y},${elem.d}`);
    fillGrid(elem);
    result++;
    char buffer[20] = {0};
    sprintf(buffer, "FILL,%d,%d,%d\n", elem.x, elem.y, elem.d);
    fwrite(buffer, 1, strlen(buffer), outFptr);
}

drawableElem *fillHeaviestArea(int x, int y) {
    drawableElem *found = NULL;
    drawableElem *elem = drawableArea;
    while (elem->next != NULL) {
        if (isInArea(x, y, *elem)) {
            elem->w = computeWeightArea(*elem);
            if (elem->w == 0) {
                elem = removeElem(elem);
            } else {
                if (found == NULL) {
                    found = elem;
                } else if (compareElem(*elem, *found) < 0) {
                    found = elem;
                }
            }
        }
        elem = elem->next;
    }
    if (found->w > 0) {
        addToResult(*found);
        elem = removeElem(found);
        return elem;
    }
    return NULL;
}

int main() {
    int rc = 0;
    if ((inFptr = fopen(INPUT_FILE, "r")) == NULL) {
        rc = -1;
        printf("Unable to open the file %s: %s\n", INPUT_FILE, strerror(errno));
        goto __error;
    }

    if ((outFptr = fopen(OUTPUT_FILE, "w")) == NULL) {
        rc = -1;
        printf("Unable to open the file %s: %s\n", OUTPUT_FILE, strerror(errno));
        goto __error;
    }

    if ((rc = parseFirstLine(inFptr)) < 0) {
        goto __error;
    }

    // printf("BaseX: %d, BaseY: %d\n", baseX, baseY);
    gridLen = baseX * baseY + 1;
    baseGrid = malloc(sizeof(char) * gridLen);
    memset(baseGrid, 0, gridLen);
    computeBaseGrid(inFptr);
    drawableArea = getDrawableArea();

    // Fill in with the heaviest area
    for (int y = 0; y < baseY; y++) {
        for (int x = 0; x < baseX; x++) {
            if (getBaseGrid(x, y) == HASH) {
                fillHeaviestArea(x, y);
            }
        }
    }

    printf("Res: %d\n", result);

    // Free memory
    free(baseGrid);
    drawableElem *elem = drawableArea;
    while (elem->next != NULL) {
        drawableArea = elem->next;
        free(elem);
        elem = drawableArea;
    }
    free(drawableArea);

    __error:
    if (inFptr) {
        if (fclose(inFptr) != 0) {
            rc = -1;
            printf("Unable to close the file %s: %s\n", INPUT_FILE, strerror(errno));
        }
    }

    if (outFptr) {
        if (fclose(outFptr) != 0) {
            rc = -1;
            printf("Unable to close the file %s: %s\n", OUTPUT_FILE, strerror(errno));
        }
    }

    return rc;
}